File: 108_palindrome-partitioning-ii/palindrome-partitioning-ii.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/palindrome-partitioning-ii
@Language: Python
@Datetime: 16-06-07 07:24
'''

class Solution:
    # @param s, a string
    # @return an integer
    def getIsPalindrome(self, s):
        isPalindrome = [[True if i == j else False\
            for j in range(len(s))] for i in range(len(s))]
        for i in range(0, len(s) - 1):
            isPalindrome[i][i + 1] = s[i] == s[i + 1]
        for l in range(2, len(s)):
            for start in range(len(s) - l):
                isPalindrome[start][start + l] =\
                    isPalindrome[start + 1][start + l - 1] and\
                    s[start] == s[start + l]
        return isPalindrome
    
    def minCut(self, s):
        if s is None or len(s) == 0:
            return 0
        f = [i for i in range(len(s) + 1)]
        isPalindrome = self.getIsPalindrome(s)
        for i in range(1, len(s) + 1):
            for j in range(i):
                if isPalindrome[j][i - 1]:
                    f[i] = min(f[i], f[j] + 1)
        return f[len(s)] - 1
                    
        

File: 108_palindrome-partitioning-ii/palindrome-partitioning-ii.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/palindrome-partitioning-ii
@Language: Java
@Datetime: 16-06-07 06:54
*/

public class Solution {
    /**
     * @param s a string
     * @return an integer
     */
    private boolean[][] getIsPalindrome(String s) {
        boolean[][] isPalindrome = new boolean[s.length()][s.length()];
        for (int i = 0; i < s.length(); i++) {
            isPalindrome[i][i] = true;
        }
        for (int i = 0; i < s.length() - 1; i++) {
            isPalindrome[i][i+1] = s.charAt(i) == s.charAt(i + 1);
        }
        for (int len = 2; len < s.length(); len++) {
            for (int start = 0; start + len < s.length(); start++) {
                isPalindrome[start][start + len] = 
                    isPalindrome[start + 1][start + len - 1] &&
                    s.charAt(start) == s.charAt(start + len);
            }
        }
        return isPalindrome;
    }
    public int minCut(String s) {
        // write your code here
        if (s == null || s.length() == 0) {
            return 0;
        }
        int[] f = new int[s.length() + 1];
        f[0] = 0;
        boolean[][] isPalindrome = getIsPalindrome(s);
        for (int i = 1; i <= s.length(); i++) {
            f[i] = i;
            for (int j = 0; j < i; j++) {
                if (isPalindrome[j][i - 1]) {
                    f[i] = Math.min(f[i], f[j] + 1);
                }
            }
        }
        return f[s.length()] - 1;
    }
};